gap/io.gi

```diff
@@ -1039,6 +1039,201 @@ function(arg)
   fi;
 end);
 
+# DIMACS format: for symmetric digraphs, one per file.
+#                Can have loops and multiple edges.
+
+InstallMethod(ReadDIMACSDigraph, "for a string", [IsString],
+function(name)
+  local file, malformed_file, int_from_string, next, split, first_char,
+  nr_vertices, vertices, vertex_labels, nr_edges, directed_edges,
+  symmetric_edges, nbs, vertex, label, i, j, digraph;
+
+  file := IO_CompressedFile(UserHomeExpand(name), "r");
+  if file = fail then
+    ErrorNoReturn("Digraphs: ReadDIMACSDigraph:\n",
+                  "cannot open the file ", name, ",");
+  fi;
+
+  # Helper function for when an error is found in the file's formatting
+  malformed_file := function()
+    IO_Close(file);
+    ErrorNoReturn("Digraphs: ReadDIMACSDigraph:\n",
+                  "the format of the file <name> cannot be understood,");
+  end;
+
+  # Helper function to read a string into a non-negative integer
+  int_from_string := function(string)
+    local int;
+    int := Int(string);
+    if int = fail or int < 0 then
+      malformed_file();
+    fi;
+    return int;
+  end;
+
+  next := IO_ReadLine(file);
+  while not IsEmpty(next) do
+    NormalizeWhitespace(next);
+
+    # the line is entirely whitespace or a comment
+    if IsEmpty(next) or next[1] = 'c' then
+      next := IO_ReadLine(file);
+      continue;
+    fi;
+
+    split := SplitString(next, " ");
+
+    # the line doesn't have a `type'
+    if Length(split[1]) <> 1 then
+      malformed_file();
+    fi;
+
+    first_char := next[1];
+
+    # digraph definition line
+    if first_char = 'p' then
+      if IsBound(vertices) or Length(split) <> 4 or split[2] <> "edge" then
+        malformed_file();
+      fi;
+      nr_vertices     := int_from_string(split[3]);
+      vertices        := [1 .. nr_vertices];
+      vertex_labels   := vertices * 0 + 1;
```

Comment:
Is this the fastest way of doing this?   Perhaps `ListWithIdenticalEntries(nr_vertices, 1)` would be better.